src/odemis/acq/milling/millmng.py

```diff
@@ -0,0 +1,229 @@
+            if align_at_milling_current:
+                self._future.running_subf = acquire([self.fib_stream])
+                data, _ = self._future.running_subf.result()
+                ref_image = data[0]
+
+            # set the milling state
+            microscope.clear_patterns()
+            microscope.set_default_patterning_beam_type(milling_channel)
+            microscope.set_high_voltage(milling_voltage, milling_channel)
+            microscope.set_beam_current(milling_current, milling_channel)
+            # microscope.set_field_of_view(milling_fov, milling_channel) # tmp: disable until matched in gui
+            microscope.set_patterning_mode(milling_mode)
+
+            # acquire a new image at the milling settings and align
+            if align_at_milling_current:
+                self._future.running_subf = acquire([self.fib_stream])
+                data, _ = self._future.running_subf.result()
+                new_image = data[0]
+                align_reference_image(ref_image, new_image, self.ion_beam)
+
+            # draw milling patterns to microscope
+            for pattern in settings.generate():
+                if isinstance(pattern, RectanglePatternParameters):
+                    microscope.create_rectangle(pattern.to_json())
+                else:
+                    raise NotImplementedError(f"Pattern {pattern} not supported") # TODO: support other patterns
+
+            # estimate the milling time
+            estimated_time = microscope.estimate_milling_time()
+            self._future.set_end_time(time.time() + estimated_time)
+
+            # start patterning (async)
+            microscope.start_milling()
+
+            # wait for milling to finish
+            elapsed_time = 0
+            wait_time = 5
+            while microscope.get_patterning_state() == "Running":
+
+                with self._future._task_lock:
+                    if self._future.cancelled() == CANCELLED:
+                        raise CancelledError()
+
+                logging.debug(f"Milling in progress... elapsed time: {elapsed_time} s, estimated time:  {estimated_time} s")
+                time.sleep(wait_time)
+                elapsed_time += wait_time
+
+        except CancelledError as ce:
+            logging.debug(f"Cancelled milling: {ce}")
+            exception = ce
+        except Exception as e:
+            logging.error(f"Error while milling: {e}")
+            exception = e
+        finally:
+            # restore imaging state
+            microscope.set_beam_current(imaging_current, milling_channel)
+            microscope.set_high_voltage(imaging_voltage, milling_channel)
+            microscope.set_field_of_view(imaging_fov, milling_channel)
+            #microscope.set_channel(milling_channel) # TODO: expose on server
+            microscope.set_active_view(2)
+            microscope.clear_patterns()
+
+            if exception:
+                raise exception
```

Comment:
You don't need that. You can `raise` (no argument) directly at the end of the exception handler, and it'll first call the `finally` block, and then raise (just like what your code does).